lib/spack/spack/compilers/__init__.py

```diff
@@ -125,6 +125,105 @@ def get_compiler_config(scope=None, init_config=True):
     return config
 
 
+def get_compiler_config_from_packages(scope=None):
+    """Return the compiler configuration from packages.yaml"""
+    packages = []
+    config = spack.config.get("packages", scope=scope)
+
+    if not config:
+        return packages
+
+    for name, entry in config.items():
+        if name not in supported_compilers() and name not in package_name_to_compiler_name.keys():
+            continue
+        packages.extend(_compiler_config_from_package_config(entry))
+
+    return packages
+
+
+def _compiler_config_from_package_config(config):
+    externals = config.get("externals", None)
+    if not externals:
+        return []
```

Comment:
I would push this as another condition to continue in the `loop` above:
```python
if "externals" not in entry:
    continue
```
and pass `entry["externals"]` as an argument to this function.